### DISCUSSION TITLE
Awaiting a function returning a never type does not raise unreachable code after the await. Is this a rust or a tokio issue?

### DISCUSSION THREAD

**Discussion by ungrichtepfl** [tokio-rs/tokio#6736] [Q&A] [answered]
Hi guys

I recently noticed that if you `await` a function returning the never type `!` the compiler (I used `rustc 1.80.0`) does not warn about unreachable code after the await, i.e. writing
```rust
pub async fn endless() -> ! {
    loop {
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
    }
}

pub async fn does_not_warn_unreachable() {
    endless().await;
    println!("Hello, world!"); // No warning
}
```
compiles without a warning, but the non-async version
```rust
pub fn endless() -> ! {
    loop {
        std::thread::sleep(std::time::Duration::from_secs(1));
    }
}

pub fn warns_unreachable() {
    endless();
    println!("Hello, world!");  // warning: unreachable statement
}
```
warns about an unreachable statement after `endless`. 

Is this a general rust `async/.await` issue with the never type or is this due to tokio?

Thanks for your help :)

**Comment by Darksonn** [comment, marked as answer]
Please file a bug with the Rust compiler instead.

**Comment by ungrichtepfl** [comment]
Moved to https://github.com/rust-lang/rust/issues/128434
